tests/unit/completion/test_completer.py

```diff
@@ -0,0 +1,189 @@
+                        cmd_utils)
+
+
+@pytest.mark.parametrize('txt, expected', [
+    (':nope|', usertypes.Completion.command),
+    (':nope |', None),
+    (':set |', usertypes.Completion.section),
+    (':set gen|', usertypes.Completion.section),
+    (':set general |', usertypes.Completion.option),
+    (':set what |', None),
+    (':set general ignore-case |', usertypes.Completion.value),
+    (':set general huh |', None),
+    (':help |', usertypes.Completion.helptopic),
+    (':quickmark-load |', usertypes.Completion.quickmark_by_name),
+    (':bookmark-load |', usertypes.Completion.bookmark_by_url),
+    (':open |', usertypes.Completion.url),
+    (':buffer |', usertypes.Completion.tab),
+    (':session-load |', usertypes.Completion.sessions),
+    (':bind |', usertypes.Completion.empty),
+    (':bind <c-x> |', usertypes.Completion.command),
+    (':bind <c-x> foo|', usertypes.Completion.command),
+    (':bind <c-x>| foo', usertypes.Completion.empty),
+    (':set| general ', usertypes.Completion.command),
+    (':|set general ', usertypes.Completion.command),
+    (':set gene|ral ignore-case', usertypes.Completion.section),
+    (':|', usertypes.Completion.command),
+    (':   |', usertypes.Completion.command),
+    (':bookmark-load      |', usertypes.Completion.bookmark_by_url),
+    ('/|', None),
+    (':open -t|', None),
+    (':open --tab|', None),
+    (':open -t |', usertypes.Completion.url),
+    (':open --tab |', usertypes.Completion.url),
+    (':open | -t', usertypes.Completion.url),
+    (':--foo --bar |', None),
+    (':tab-detach |', None),
+    (':bind --mode=caret <c-x> |', usertypes.Completion.command),
+    #(':bind --mode caret <c-x> |', usertypes.Completion.command), KNOWN BUG
+    (':set -t -p |', usertypes.Completion.section),
+    (':open -- |', None),
+])
+def test_update_completion(txt, expected, cmd, completer_obj,
+                           completion_widget_stub):
+    """Test setting the completion widget's model based on command text."""
+    # this test uses | as a placeholder for the current cursor position
+    cursor_pos = txt.index('|')
+    cmd.setText(txt.replace('|', ''))
+    cmd.setCursorPosition(cursor_pos)
+    completer_obj.update_completion()
+    if expected is None:
+        assert not completion_widget_stub.set_model.called
+    else:
+        assert completion_widget_stub.set_model.call_count == 1
+        arg = completion_widget_stub.set_model.call_args[0][0]
+        # the outer model is just for sorting; srcmodel is the completion model
+        assert arg.srcmodel.kind == expected
+
+def test_completion_item_prev(completer_obj, cmd, completion_widget_stub,
+                              config_stub):
+    """Test that completion_item_prev emits next_prev_item."""
+    cmd.setText(':')
+    slot = unittest.mock.Mock()
+    completer_obj.next_prev_item.connect(slot)
+    completer_obj.completion_item_prev()
+    slot.assert_called_with(True)
```

Comment:
I'd use [`qtbot.waitSignal`](http://pytest-qt.readthedocs.io/en/stable/reference.html#pytestqt.qtbot.QtBot.waitSignal) for this kind of thing - something like:

```python
with qtbot.waitSignal(completer_obj.next_prev_item) as blocker:
    completer_obj.completion_item_prev()
assert blocker.args == [True]
```